sdk/python/kfp_tekton/compiler/_op_to_template.py

```diff
@@ -208,68 +216,128 @@ def _op_to_template(op: BaseOp):
+                        args = []
+                        for a in s['args']:
+                            if path in a:
+                                a = a.replace(path, '$(results.%s.path)' % name)
+                                need_copy_step = False
+                            args.append(a)
+                        s['args'] = args
+                # If file output path cannot be found/replaced, use emptyDir to copy it to the tekton/results path
+                if need_copy_step:
+                    copy_results_step['script'] = copy_results_step['script'] + 'cp ' + path + ' $(results.%s.path);' % name + '\n'
+                    mountPath = path.rsplit("/", 1)[0]
+                    if mountPath not in mounted_param_paths:
+                        volume_mount_step_template.append({'name': name, 'mountPath': path.rsplit("/", 1)[0]})
+                        volume_template.append({'name': name, 'emptyDir': {}})
+                        mounted_param_paths.append(mountPath)
+                # Record what artifacts are moved to result parameters.
+                parameter_name = (processed_op.name + '-' + name).replace(' ', '-').replace('_', '-')
+                replaced_param_list.append(parameter_name)
+        copy_artifacts_step = {}
+        if outputs_dict.get('artifacts'):
+            """
+            For storing artifacts, we will be using the minio/mc image because we need to upload artifacts to any type of
+            object storage and endpoint. The minio/mc is the best image suited for this task because the default KFP
+            is using minio and it also works well with other s3/gcs type of storage. 
+
+            - image: minio/mc
+              name: copy-artifacts
+              script: |
+                  #!/usr/bin/env sh
+                  mc config host add storage http://minio-service.$NAMESPACE:9000 $AWS_ACCESS_KEY_ID $AWS_SECRET_ACCESS_KEY
+                  mc cp /tmp/file.txt storage/$(inputs.params.bucket)/runs/$PIPELINERUN/$PODNAME/file.txt
+            """
+            # TODO: Pull default values from KFP configmap when integrated with KFP.
+            storage_location = outputs_dict['artifacts'][0].get('s3', {})
+            insecure = storage_location.get("insecure", True)
+            endpoint = storage_location.get("endpoint", "minio-service.$NAMESPACE:9000")
+            # We want to use the insecure flag to figure out whether to use http or https scheme
+            endpoint = re.sub(r"https?://", "", endpoint)
+            endpoint = 'http://' + endpoint if insecure else 'https://' + endpoint
+            access_key = storage_location.get("accessKeySecret", {"name": "mlpipeline-minio-artifact", "key": "accesskey"})
+            secret_access_key = storage_location.get("secretKeySecret", {"name": "mlpipeline-minio-artifact", "key": "secretkey"})
+            bucket = storage_location.get("bucket", "mlpipeline")
+            copy_artifacts_step = {
+                'image': 'minio/mc',
```

Comment:
this block is growing really large, should we move this into a separate funtion/method and maybe we could make the `copy_artifacts_step` into a Python template string. In the future we could keep all the highly descriptive code snippets (YAML with a few parameters) in a separate folder as opposed to mixing it with the Python code. We could track that in a separate issue :-)

Reply:
I agree, there are big chunk of code templates introduced for resourceOp and artifacts because we need to use an extra step/task to support these additional features. There's still some merging happening for volume and volume snapshot op along with some new changes to resource op that Vincent did yesterday. I can open an issue to refactor this file and compiler.py.